File: etc/gh/basic_tree.py
class TreeNode:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self, root_data):
        self.root = TreeNode(root_data)

    def insert_left(self, current_node, data):
        if current_node.left is None:
            current_node.left = TreeNode(data)
        else:
            new_node = TreeNode(data)
            new_node.left = current_node.left
            current_node.left = new_node

    def insert_right(self, current_node, data):
        if current_node.right is None:
            current_node.right = TreeNode(data)
        else:
            new_node = TreeNode(data)
            new_node.right = current_node.right
            current_node.right = new_node

    def inorder_traversal(self, node):
        if node is not None:
            self.inorder_traversal(node.left)
            print(node.data, end=' ')
            self.inorder_traversal(node.right)

    def preorder_traversal(self, node):
        if node is not None:
            print(node.data, end=' ')
            self.preorder_traversal(node.left)
            self.preorder_traversal(node.right)

    def postorder_traversal(self, node):
        if node is not None:
            self.postorder_traversal(node.left)
            self.postorder_traversal(node.right)
            print(node.data, end=' ')

# 사용 예시
if __name__ == "__main__":
    bt = BinaryTree(1)
    bt.insert_left(bt.root, 2)
    bt.insert_right(bt.root, 3)
    bt.insert_left(bt.root.left, 4)
    bt.insert_right(bt.root.left, 5)

    print("Inorder traversal:")
    bt.inorder_traversal(bt.root)  # 출력: 4 2 5 1 3

    print("\nPreorder traversal:")
    bt.preorder_traversal(bt.root)  # 출력: 1 2 4 5 3

    print("\nPostorder traversal:")
    bt.postorder_traversal(bt.root)  # 출력: 4 5 2 3 1

File: etc/gh/data_structure/basic_tree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct TreeNode {
    int data;
    struct TreeNode* left;
    struct TreeNode* right;
} TreeNode;

TreeNode* create_node(int data) {
    TreeNode* new_node = (TreeNode*)malloc(sizeof(TreeNode));
    new_node->data = data;
    new_node->left = NULL;
    new_node->right = NULL;
    return new_node;
}

void insert_left(TreeNode* current_node, int data) {
    if (current_node->left == NULL) {
        current_node->left = create_node(data);
    } else {
        TreeNode* new_node = create_node(data);
        new_node->left = current_node->left;
        current_node->left = new_node;
    }
}

void insert_right(TreeNode* current_node, int data) {
    if (current_node->right == NULL) {
        current_node->right = create_node(data);
    } else {
        TreeNode* new_node = create_node(data);
        new_node->right = current_node->right;
        current_node->right = new_node;
    }
}

void inorder_traversal(TreeNode* node) {
    if (node != NULL) {
        inorder_traversal(node->left);
        printf("%d ", node->data);
        inorder_traversal(node->right);
    }
}

void preorder_traversal(TreeNode* node) {
    if (node != NULL) {
        printf("%d ", node->data);
        preorder_traversal(node->left);
        preorder_traversal(node->right);
    }
}

void postorder_traversal(TreeNode* node) {
    if (node != NULL) {
        postorder_traversal(node->left);
        postorder_traversal(node->right);
        printf("%d ", node->data);
    }
}

int main() {
    TreeNode* root = create_node(1);
    insert_left(root, 2);
    insert_right(root, 3);
    insert_left(root->left, 4);
    insert_right(root->left, 5);

    printf("Inorder traversal:\n");
    inorder_traversal(root);  // 출력: 4 2 5 1 3
    printf("\n");

    printf("Preorder traversal:\n");
    preorder_traversal(root);  // 출력: 1 2 4 5 3
    printf("\n");

    printf("Postorder traversal:\n");
    postorder_traversal(root);  // 출력: 4 5 2 3 1
    printf("\n");

    return 0;
}
